ipa-core/src/protocol/ipa_prf/prf_eval.rs

```diff
@@ -74,39 +97,45 @@ where
 pub async fn eval_dy_prf<C, const N: usize>(
     ctx: C,
     record_id: RecordId,
-    k: &AdditiveShare<Fp25519>,
+    key: &AdditiveShare<Fp25519>,
     x: AdditiveShare<Fp25519, N>,
 ) -> Result<[u64; N], Error>
 where
-    C: Context,
-    Fp25519: Vectorizable<N>,
-    RP25519: Vectorizable<N, Array = StdArray<RP25519, N>>,
-    AdditiveShare<Fp25519, N>: SecureMul<C> + FromPrss,
-    StdArray<RP25519, N>: Sendable,
+    C: UpgradedContext<Field = Fp25519>,
+    AdditiveShare<Fp25519, N>: PrfSharing<C, N, Field = Fp25519>,
+    Fp25519: FieldSimd<N>,
+    RP25519: Vectorizable<N>,
 {
-    let sh_r: AdditiveShare<Fp25519, N> =
-        ctx.narrow(&Step::GenRandomMask).prss().generate(record_id);
+    // compute x+k (and upgrade the share)
+    let y = (x + key.expand())
+        .upgrade(ctx.narrow(&Step::UpgradeY), record_id)
+        .await?;
 
-    //compute x+k
-    let mut y = x + k.expand();
+    let r: AdditiveShare<Fp25519, N> = ctx.narrow(&Step::GenRandomMask).prss().generate(record_id);
+    // compute (g^left, g^right).
+    // Note that it does not get upgraded because this
+    // value is revealed immediately without any multiplications
+    let sh_gr = AdditiveShare::<RP25519, N>::from(r.clone());
 
-    //compute y <- r*y
-    y = y
-        .multiply(&sh_r, ctx.narrow(&Step::MultMaskWithPRFInput), record_id)
+    // upgrade r and compute y <- r*y
+    let r = r.upgrade(ctx.narrow(&Step::UpgradeMask), record_id).await?;
+    let y = y
+        .multiply(&r, ctx.narrow(&Step::MultMaskWithPRFInput), record_id)
         .await?;
 
-    //compute (g^left, g^right)
-    let sh_gr = AdditiveShare::<RP25519, N>::from(sh_r);
-
-    //reconstruct (z,R)
-    // TODO: these should invoke reveal via the trait when this function
-    // takes a context of an appropriate type.
-    let gr = malicious_reveal(ctx.narrow(&Step::RevealR), record_id, None, &sh_gr)
-        .await?
-        .unwrap();
-    let z = malicious_reveal(ctx.narrow(&Step::Revealz), record_id, None, &y)
-        .await?
-        .unwrap();
+    // validate everything before reveal
+    ctx.validate_record(record_id).await?;
+    let (gr, z): (
+        <RP25519 as Vectorizable<N>>::Array,
+        <Fp25519 as Vectorizable<N>>::Array,
+    ) = try_join(
+        // TODO: these should invoke reveal via the trait when this function
+        // takes a context of an appropriate type.
+        malicious_reveal(ctx.narrow(&Step::RevealR), record_id, None, &sh_gr)
+            .map(|v| v.map(|arr| arr.unwrap())),
```

Comment:
You will need the changes in #1229 for this to work (yet another piece of #1112), but I think it would be cleaner and more correct to do:
```suggestion
        reveal(ctx.narrow(&Step::RevealR), record_id, &sh_gr),
```
The behavior will be different when running against a semi-honest context. It's not harmful to do the malicious reveal when the context is semi-honest, but I think it's better to follow the principle of not invoking primitives (`semi_honest_reveal` / `malicious_reveal`) directly from high-level protocols that use a generic context. (The opposite case, invoking `semi_honest_reveal`, would be harmful, since it may forgo a malicious reveal when one is called for.)